restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -0,0 +1,239 @@
+import io.reactivex.Single;
+import io.stargate.db.ImmutableParameters;
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.datastore.ResultSet;
+import io.stargate.db.datastore.Row;
+import io.stargate.db.query.builder.AbstractBound;
+import io.stargate.db.query.builder.BuiltSelect;
+import io.stargate.db.schema.Column;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+/** Executes pre-built document queries, groups document rows and manages document pagination. */
+public class QueryExecutor {
+  private final Accumulator TERM = new Accumulator();
+
+  private final DataStore dataStore;
+
+  public QueryExecutor(DataStore dataStore) {
+    this.dataStore = dataStore;
+  }
+
+  public Flowable<RawDocument> queryDocs(
+      AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    return queryDocs(1, query, pageSize, pagingState);
+  }
+
+  public Flowable<RawDocument> queryDocs(
+      int identityDepth, AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    BuiltSelect select = (BuiltSelect) query.source().query();
+    if (identityDepth < 1 || identityDepth > select.table().primaryKeyColumns().size()) {
+      throw new IllegalArgumentException("Invalid document identity depth: " + identityDepth);
+    }
+
+    List<Column> idColumns = select.table().primaryKeyColumns().subList(0, identityDepth);
+
+    return execute(query, pageSize, pagingState)
+        .concatMap(rs -> Flowable.fromIterable(seeds(rs, idColumns)), 1)
+        .concatWith(Single.just(TERM))
+        .scan(Accumulator::combine)
+        .filter(Accumulator::isComplete)
+        .map(Accumulator::toDoc);
+  }
+
+  public Flowable<ResultSet> execute(AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    return fetchPage(query, pageSize, pagingState)
+        .compose( // Expand BREADTH_FIRST to reduce the number of "proactive" page requests
+            FlowableTransformers.expand(
+                rs -> fetchNext(rs, pageSize, query), ExpandStrategy.BREADTH_FIRST, 1));
+  }
+
+  private Flowable<ResultSet> fetchPage(
+      AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    CompletableFuture<ResultSet> futureResult = new CompletableFuture<>();
+    return Flowable.create(
+            (FlowableOnSubscribe<ResultSet>)
+                emitter -> // separate subscription from execution - see doOnRequest
+                futureResult.whenComplete(
+                        (rows, t) -> {
+                          if (t != null) {
+                            emitter.onError(t);
+                          } else {
+                            emitter.onNext(rows);
+                            emitter.onComplete();
+                          }
+                        }),
+            BackpressureStrategy.BUFFER)
+        .doOnRequest( // execute only when requested
```

Comment:
I can remove this, but relying on subscription callbacks will cause more "proactive" page fetching. The RxJava framework code generally subscribes to the next element whenever it emits the current element, so we'd always have some "proactive" page requests.

I tested this as part of my local code evolution and found that relying purely on subscription callbacks leads to one extra "proactive" page request compared to what we have now. So now we have 2 pages fetched at a minimum, with subscription-driven execution we will have 3 pages min.

This may or may not matter because high-level pipelines generally have `.limit()` operators. I did not investigate it at that level of details. Perhaps I should :)

Reply:
Yea I think there is no better way of doing it atm..